src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala

```diff
@@ -154,7 +154,67 @@ class PostgresDirectoryDAO(protected val dbRef: DbReference,
     */
   override def removeGroupMember(groupId: WorkbenchGroupIdentity, removeMember: WorkbenchSubject): IO[Boolean] = ???
 
-  override def isGroupMember(groupId: WorkbenchGroupIdentity, member: WorkbenchSubject): IO[Boolean] = ???
+  override def isGroupMember(groupId: WorkbenchGroupIdentity, member: WorkbenchSubject): IO[Boolean] = {
+    val gm = GroupMemberTable.syntax("gm")
+
+    val memberClause: SQLSyntax = member match {
+      case WorkbenchGroupName(groupName) =>
+        val g = GroupTable.syntax("g")
+        sqls"sg.member_group_id = (select ${g.id} from ${GroupTable.as(g)} where ${g.name} = $groupName)"
+
+      case FullyQualifiedPolicyId(FullyQualifiedResourceId(resourceTypeName, resourceId), accessPolicyName) =>
+        val rt = ResourceTypeTable.syntax("rt")
+        val r = ResourceTable.syntax("r")
+        val p = PolicyTable.syntax("p")
+        sqls"""sg.member_group_id = (select ${p.groupId}
+              from ${ResourceTypeTable.as(rt)}
+              join ${ResourceTable.as(r)} on ${rt.id} = ${r.resourceTypeId}
+              join ${PolicyTable.as(p)} on ${r.id} = ${p.resourceId}
+              where ${rt.resourceTypeName} = $resourceTypeName)
+              and ${r.name} = $resourceId
+              and ${p.name} = $accessPolicyName)"""
+
+      case WorkbenchUserId(userId) => sqls"sg.member_user_id = $userId"
+
+      case _ => throw new WorkbenchException(s"illegal member $member")
+    }
+
+    val topGroupQuery = groupId match {
+      case WorkbenchGroupName(groupName) =>
+        val g = GroupTable.syntax("g")
+        sqls"select ${g.id}, ${gm.memberGroupId}, ${gm.memberUserId} from ${GroupTable.as(g)} join ${GroupMemberTable.as(gm)} on ${g.id} = ${gm.groupId} WHERE ${g.name} = $groupName"
+
+      case FullyQualifiedPolicyId(FullyQualifiedResourceId(resourceTypeName, resourceId), accessPolicyName) =>
+        val rt = ResourceTypeTable.syntax("rt")
+        val r = ResourceTable.syntax("r")
+        val p = PolicyTable.syntax("p")
+        sqls"""select ${p.groupId}, ${gm.memberGroupId}, ${gm.memberUserId}
+              from ${ResourceTypeTable.as(rt)}
+              join ${ResourceTable.as(r)} on ${rt.id} = ${r.resourceTypeId}
+              join ${PolicyTable.as(p)} on ${r.id} = ${p.resourceId}
+              join ${GroupMemberTable.as(gm)} on ${p.groupId} = ${gm.groupId}
+              where ${rt.resourceTypeName} = $resourceTypeName)
+              and ${r.name} = $resourceId
+              and ${p.name} = $accessPolicyName"""
+
+    }
+
+    runInTransaction { implicit session =>
+      // https://www.postgresql.org/docs/9.6/queries-with.html
+      // in the recursive query below, UNION, as opposed to UNION ALL, should break out of cycles because it removes duplicates
+      val query = sql"""WITH RECURSIVE sub_group(parent_group_id, member_group_id, member_user_id) AS (
```

Comment:
Might want to extract all of this up to constructing `query` into its own method, something like `traceGroupMembership`.  I still want to read up on `WITH RECURSIVE`, but if I'm reading this right, we'll end up with a row of data that represents the path of group memberships between some parent group and the member.  Might be useful to be able to look at that in some cases?  

Then your `isMemberOf` method is really just the `sql("(${traceGroupMembership(groupId, member)}) SELECT count(1)...")`

Or ignore me....I just love extracting methods!

Reply:
I expect we will end up extracting a bunch but we need to dig into the other queries to see how to do it best